command/recorderd/proofer.go

```diff
@@ -7,22 +7,141 @@ package main
 import (
 	"encoding/binary"
 	"encoding/json"
+	"fmt"
+	"math"
+	"runtime"
+	"sync"
 	"sync/atomic"
 	"time"
 
 	zmq "github.com/pebbe/zmq4"
 
 	"github.com/bitmark-inc/bitmarkd/blockdigest"
 	"github.com/bitmark-inc/bitmarkd/blockrecord"
+	"github.com/bitmark-inc/exitwithstatus"
 	"github.com/bitmark-inc/logger"
 )
 
 const (
-	proofRequest = "inproc://blocks.request"  // to fair-queue block requests
-	dispatch     = "inproc://blocks.dispatch" // proofer fetches from here
+	proofRequest        = "inproc://blocks.request"  // to fair-queue block requests
+	dispatch            = "inproc://blocks.dispatch" // proofer fetches from here
+	errorProoferID      = -1
+	ProoferLoggerPrefix = "proofer"
 )
 
-var proofQueueDepth uint64
+var (
+	proofQueueDepth uint64
+)
+
+type Proofer interface {
+	startHashing()
+	stopHashing()
+	isWorking() bool
+	refresh()
+}
+
+type ProoferData struct {
+	sync.RWMutex
+	activeThreadCount uint32
+	prevThreadCount   uint32
+	proofIDs          []bool
+	stopChannel       chan struct{}
+	log               *logger.L
+	workingNow        bool
+	cpuCount          int
+	reader            *ConfigReader
+}
+
+func newProofer(log *logger.L, reader *ConfigReader) Proofer {
+	cpuCount := runtime.NumCPU()
+	return &ProoferData{
+		log:         log,
+		proofIDs:    make([]bool, cpuCount),
+		stopChannel: make(chan struct{}, cpuCount),
+		cpuCount:    cpuCount,
+		workingNow:  true,
+		reader:      reader,
+	}
+}
+
+func (p *ProoferData) startHashing() {
+	p.log.Infof("receive start hashing request, current active thread %d",
+		p.activeThread())
+	p.setWorking(true)
+	if p.activeThread() < 1 {
+		p.createProofer(p.reader.optimalThreadCount())
+	}
+}
+
+func (p *ProoferData) stopHashing() {
+	p.log.Infof("receive stop hashing request, current active thread %d",
+		p.activeThread())
+	p.setWorking(false)
+	p.deleteProofer(int32(p.activeThread()))
+}
+
+func (p *ProoferData) deleteProofer(count int32) {
+	p.log.Infof("delete %d goroutine from hashing", count)
+	for i := int32(0); i < count; i++ {
+		p.log.Debug("send signal to stop channel")
+		p.stopChannel <- struct{}{}
+	}
+}
+
+func (p *ProoferData) isWorking() bool {
+	return p.workingNow
+}
+
+func (p *ProoferData) setWorking(working bool) {
+	p.workingNow = working
+}
+
+func (p *ProoferData) refresh() {
+	if !p.changed() || !p.isWorking() {
+		return
+	}
+
+	increment := p.differenceToTargetThreadCount(
+		p.reader.optimalThreadCount(),
+		p.activeThread(),
+	)
+
+	p.log.Infof("refresh settings, active goroutine %d, need to increase %d",
+		p.activeThread(), increment)
+
+	if increment > 0 {
+		p.createProofer(uint32(increment))
+		return
+	}
+	p.deleteProofer(-increment)
+}
+
+func (p *ProoferData) changed() bool {
+	return p.prevThreadCount != p.reader.optimalThreadCount()
```

Comment:
1. In my opinion, config reader and file watcher is different, reader only reads config, watcher only watcher for file change.

2. The info can be found in log file, it prints out every scheduled time slice

3. I check the dictionary, optimal and optimum are same meaning, use which one should have no difference, so i guess it would be easier to keep the original one